Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php

```diff
@@ -87,6 +90,51 @@ static function _checkGoto($a_target)
 
 		return false;
 	}
+
+	/**
+	 * Get operators
+	 */
+	public static function getConditionOperators()
+	{
+		include_once './Services/AccessControl/classes/class.ilConditionHandler.php';
+		return array(
+			ilConditionHandler::OPERATOR_ACCREDITED_OR_PASSED
+		);
+	}
+
+	/**
+	 *
+	 * @global ilObjUser $ilUser
+	 * @param type $a_obj_id
+	 * @param type $a_operator
+	 * @param type $a_value
+	 * @param type $a_usr_id
+	 * @return boolean
+	 */
+	public static function checkCondition($a_obj_id,$a_operator,$a_value,$a_usr_id)
+	{
+		include_once './Services/AccessControl/classes/class.ilConditionHandler.php';
+		include_once "./Modules/StudyProgramme/classes/class.ilStudyProgrammeUserProgress.php";
+
+		switch($a_operator) {
+			case ilConditionHandler::OPERATOR_ACCREDITED_OR_PASSED:
+				$valid_progress = array(
+					ilStudyProgrammeProgress::STATUS_COMPLETED,
+					ilStudyProgrammeProgress::STATUS_ACCREDITED
+				);
+
+				$prg_user_progress = ilStudyProgrammeUserProgress::getInstancesForUser($a_obj_id, $a_usr_id);
+				foreach ($prg_user_progress as $progress) {
+					if( in_array($progress->getStatus(), $valid_progress)) {
+						return true;
+					}
+				}
```

Comment:
Let's take the time for a little refactoring in the SP here:

* Move the static-functions on `ilStudyProgrammeUserProgress` that query from the database to a new non static `ilStudyProgrammeUserProgressDB`.
* Look for usage sites of these static functions. Replace them by a `$this->getStudyProgrammeUserProgressDB()->...`.
* If the object using the static function is one where we control the creation ourselves, pass the new DB as a dependency in the constructor and implement `getStudyProgrammeUserProgressDB` accordingly. If not, implement `getStudyProgrammeUserProgressDB` by creating and caching an instance. If a static function uses the statics, implement `getStudyProgrammeUserProgressDB` as static as well and call it as `static::getStudyProgrammeUserProgressDB`. This will allow to overwrite the method for a test.
* This should open some opportunities for tests. Write them.
* This will make it possible to test this very method. Write the test.